File: python/Combination-Sum/Combination-Sum.py
#! /usr/bin/env python
#! -*- coding=utf-8 -*-
# Date: 2019-11-30
# Author: Bryce


class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        res = []
        out = []
        self.DFS(candidates, target, 0, out, res)
        return res
    def DFS(self, candidates: List[int], target: int, start: int, out: List[int], res: List[List[int]]):
        if target < 0: return
        if target == 0:
            res.append(out[:])
            return
        n = len(candidates)
        for i in range (start,n):
            out.append(candidates[i])
            self.DFS(candidates,target-candidates[i],i,out,res)
            out.pop()  ## out.pop(-1) or del out[-1]

File: cpp/Combination-Sum/Combination-Sum.cpp
/// Source : https://leetcode-cn.com/problems/combination-sum/
/// Author : bryce
/// Time   : 2019-11-30


#include <iostream>
#include <vector>
using namespace std;
class Solution {
public:
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<vector<int>> res;
        vector<int> out;
        if (candidates.empty())
            return res;
        DFS(candidates, target, 0, out, res);
        return res;
    }
    
    void DFS(vector<int>& candidates, int target, int start, vector<int>& out, vector<vector<int>>& res)
    {
        if (target < 0) return;
        if (target == 0)
        {
            res.push_back(out);
            return;
        }
        int n = candidates.size();
        for (int i = start; i<n; i++)
        {
            out.push_back(candidates[i]);
            DFS(candidates, target - candidates[i], i, out, res);
            out.pop_back();
        }
    }
};

int main()
{
    int target = 7;
    vector<int> candidates = {2,3,6,7};
    vector<vector<int>> res = Solution().combinationSum(candidates,target);
    cout << res[0][0]<<endl;
    system("pause");
    return 0;
}
